### ISSUE TITLE
Борьба с переобучением

### ISSUE THREAD

**Issue by igsha** [13Fire13nyu/MasterDegree#10]
Наблюдается переобучение модели
![image](https://user-images.githubusercontent.com/5345170/168422586-050d3460-15ed-4e4e-84fe-6a0fc3aaa816.png)

Рассмотреть 2 способа борьбы (итоговая точность при это не должна падать):
* с помощью регуляризации на некоторых слоях https://keras.io/api/layers/regularizers/
* с помощью введения `Dropout` слоя перед выходным слоем

Вдохновляться лабораторными работами. Там это рассмотрено.

**Comment by 13Fire13nyu** [comment]
kernel_regularizer: Регулятор для применения штрафа к ядру слоя
bias_regularizer: Регулятор для применения штрафа к смещению слоя.
activity_regularizer: Регулятор для применения штрафа к выходным данным слоя.
Можете подробнее рассказать об этих пунктах регуляризации? Я искала какую-то подробную информацию, но не очень в этом преуспела. Есть ли какие-то формулы, в которых применяются такие регуляторы? Я помню только стандартные формулы L1 и L2 регуляризаций, где просто использовались веса множителем - лямбдой.

**Comment by igsha** [comment]
Это и есть регуляризация, которая описана в книже Умняшкина "Распознавание образов". Используется на этапе обратного распространения ошибок. Нужно просто вставить в слои и посмотреть на результат. Точность может понизиться, но для этого нужно тогда повысить количество эпох. Если точность так и не повысилась, тогда этот метод нам не подходит.
Попробуйте применить просто к весам (`kernel_regularizer`) какой-нибудь l2-регуляризатор. И вообще к чему возникают вопросы? Уже есть какие-то ошибки и что-то не получается?

**Comment by 13Fire13nyu** [comment]
> Это и есть регуляризация, которая описана в книже Умняшкина "Распознавание образов". Используется на этапе обратного распространения ошибок. Нужно просто вставить в слои и посмотреть на результат. Точность может понизиться, но для этого нужно тогда повысить количество эпох. Если точность так и не повысилась, тогда этот метод нам не подходит. Попробуйте применить просто к весам (`kernel_regularizer`) какой-нибудь l2-регуляризатор. И вообще к чему возникают вопросы? Уже есть какие-то ошибки и что-то не получается?

Я уже рассмотрела изменение параметра rate в слое Dropout. Лучший результат - 92% при rate=0.6, но переобучение всё ещё есть, хоть и стало меньше. Сейчас запускаю код для различных вариаций применения этой регуляризации с учётом слоя Dropout при rate=0.6, конкретно следующие:
1. L1 - регуляризации для ядра и смещения;
2. L1 - регуляризации для активации;
3. L2 - регуляризации для ядра и смещения;
4. L2 - регуляризация для активации. 
Пока проверила первые 2 пункта для 100 эпох при изменении параметра lambda. Для первого пункта добилась точности 93% при lambda=1e-6. Для 2 пункта точность стала ниже. Сами графики на наличие переобучения ещё не рассматривала. 
Регуляризацию применяю на всех трёх скрытых слоях. 
Задалась вопросом об информации об используемых регуляризации, потому что надо будет потом это в дипломе описывать.

**Comment by igsha** [comment]
Dropout 0.6 - это очень много. Это отбрасывание 0.6 весов нейросети. Возможно есть проблема с входными данными.

Параметры регуляризации подбирать нужно с помощью кросс-валидации. Регуляризация для активации не нужна. Для смещений скорее всего тоже - это нужно понять при анализе влияния параметров на переобучение.

**Comment by 13Fire13nyu** [comment]
Возникли проблемы с кросс-валидацией.
```
from tensorflow.keras import regularizers

for spectrogram, _ in spectrogram_ds.take(1):
    input_shape = spectrogram.shape
print('Input shape:', input_shape)

def create_model_1(lamb=1e-06):
    inputs = layers.Input(shape = input_shape, name = "inputs")
    x = layers.Flatten()(inputs)
    x = layers.Dense(512, activation = 'relu',
                     kernel_regularizer=regularizers.l2(lamb),
                     name = "relu_1")(x)
    x = layers.Dense(512, activation = 'relu',
                     kernel_regularizer=regularizers.l2(lamb), 
                     name = "relu_2")(x)
    x = layers.Dense(512, activation = 'relu',
                     kernel_regularizer=regularizers.l2(lamb),
                     name = "relu_3")(x)
    outputs = layers.Dense(12,activation = 'softmax', name = "dense_end")(x)
    model = tf.keras.Model(inputs = inputs, outputs = outputs)
    model.compile(optimizer='adam',
                  loss = 'sparse_categorical_crossentropy',
                  metrics=['accuracy'])
    return model
```
```
from sklearn.model_selection import GridSearchCV

EPOCHS = 100
parameters = {'lamb': np.logspace(-6, 2, 8)}

model = create_model_1()
model_clf = GridSearchCV(model,parameters,cv=10,n_jobs=-1)
history = model_clf.fit(train_ds, validation_data=val_ds,epochs=EPOCHS, verbose=0,
                    #callbacks=tf.keras.callbacks.EarlyStopping(verbose=1, patience=2),
                    )
for param_name in sorted(parameters.keya()):
    print("%s: %r" % (param_name, model_clf.best_params_[param_nema]))
print("Best score: %.3f" % model_clf.best_score_)
```
![image](https://user-images.githubusercontent.com/47567199/169134171-28051975-9ece-4052-8ae6-fc1c3e1196bc.png)

**Comment by igsha** [comment]
`GridSearchCV` невозможно использовать для keras-моделей. Нужна обёртка `keras.wrappers.scikit_learn.KerasClassifier`: https://www.kaggle.com/code/shujunge/gridsearchcv-with-keras/notebook

**Comment by 13Fire13nyu** [comment]
> `GridSearchCV` невозможно использовать для keras-моделей. Нужна обёртка `keras.wrappers.scikit_learn.KerasClassifier`: https://www.kaggle.com/code/shujunge/gridsearchcv-with-keras/notebook

Все еще есть ошибка:
```
%%time
from sklearn.model_selection import GridSearchCV
from keras.wrappers.scikit_learn import KerasClassifier

EPOCHS = 100
lm = np.logspace(-6, 2, 8)
param_grid = dict(lamb = lm)

model = KerasClassifier(build_fn=create_model_1)
grid = GridSearchCV(estimator=model, param_grid=param_grid)    
history = model_clf.fit(train_ds, validation_data=val_ds, epochs=EPOCHS, verbose=0)
```
![image](https://user-images.githubusercontent.com/47567199/169349597-27910900-da6c-4bd7-956a-2fb5f291795d.png)

**Comment by 13Fire13nyu** [comment]
> `GridSearchCV` невозможно использовать для keras-моделей. Нужна обёртка `keras.wrappers.scikit_learn.KerasClassifier`: https://www.kaggle.com/code/shujunge/gridsearchcv-with-keras/notebook

```

%%time
from sklearn.model_selection import GridSearchCV
from keras.wrappers.scikit_learn import KerasClassifier

EPOCHS = 100
lm = np.logspace(-6, 2, 8)
param_grid = dict(lamb = lm)

model = KerasClassifier(build_fn=create_model_1)
grid = GridSearchCV(estimator=model, param_grid=param_grid)    
history = grid.fit(train_ds, validation_data=val_ds, epochs=EPOCHS, verbose=0)
```
![image](https://user-images.githubusercontent.com/47567199/169350020-13f4c4ea-dd8b-4452-9b9d-74fd7c6f7a64.png)

**Comment by igsha** [comment]
Модели нужно подавать массивы данных, как это делалось у на лабораторных. У Вас же используется отложенный датасет. Это особенность tensorflow. Сам tensorflow может и без датасета работать, напрямую с данными. Но у Вас в коде почему-то изначально использовался сложный подход с датасетом.

Можете пока отложить эту задачу, так как я уже писал, возможно проблема в размере данных. Есть вероятность, что после #12 переобучение само уйдёт.

**Comment by 13Fire13nyu** [comment]
Мне нужно будет делать эту задачу с учетом того, что подбор шага окна и количества спектрограмм устранил переобучение?

**Comment by igsha** [comment]
Пока её откладываем, так как переобучение устранилось. Возможно потом снова вернёмся, чтобы улучшить характеристики.

**Comment by igsha** [comment]
Скорее всего это будет у каждой модели. Поэтому способы борьбы будут отличаться. Выпишу в отдельных задачах.
